bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPathDirective.kt

```diff
@@ -0,0 +1,123 @@
+package com.bugsnag.android
+
+/**
+ * A directive determines how document path commands are interpreted. Making a container,
+ * getting from a container, or setting to a container will require different operations to
+ * complete, depending on the context.
+ */
+interface DocumentPathDirective {
+    /**
+     * Make a new container at the current level.
+     *
+     * @return The new container.
+     */
+    fun newContainer(): Any
+
+    /**
+     * Get an object from the container at the current level.
+     *
+     * @param container The container to get the object from.
+     * @return The resulting object, or null if not found.
+     */
+    fun getFromContainer(container: Any): Any?
+
+    /**
+     * Set an object in a container at the current level.
+     *
+     * @param container The container to set a value in.
+     * @param value The value to set (can be null, which usually means "remove")
+     */
+    fun setInContainer(container: Any, value: Any?)
+
+    /**
+     * Modifies a particular key in a map, or creates a new (String, Object) map.
+     * Setting null removes the value.
+     */
+    class MapKeyDirective(private val key: String) : DocumentPathDirective {
+        override fun newContainer(): Any {
+            return mutableMapOf<String, Any>()
+        }
+
+        override fun getFromContainer(container: Any): Any? {
+            return (container as Map<*, *>)[key]
+        }
+
+        override fun setInContainer(container: Any, value: Any?) {
+            @Suppress("UNCHECKED_CAST")
+            val asMap = container as MutableMap<String, Any>
+            if (value == null) {
+                asMap.remove(
+                    key
+                )
+            } else {
+                asMap[key] = value
+            }
+        }
+    }
+
+    /**
+     * Modifies a list at the current level, or creates a new list.
+     * Setting null removes the item at the specified index.
+     */
+    open class ListIndexDirective(private val index: Int) : DocumentPathDirective {
+        override fun newContainer(): Any {
+            return mutableListOf<Any>()
+        }
+
+        override fun getFromContainer(container: Any): Any? {
+            return (container as List<Any?>)[index]
```

Comment:
If an index other than `-1` was specified in the path, it's expected that the list already exists in the document and has at least that many items in it already before running the directive. The most common case will be to use `-1` rather than an index, unless you're absolutely SURE of what the list will be populated with.